apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/block-iframe-preview.js

```diff
@@ -0,0 +1,206 @@
+				each( attributes, ( { name, value } ) => ( node[ name ] = value ) );
+
+				if ( innerHTML ) {
+					node.innerHTML = innerHTML;
+				}
+
+				iFrameDomReferences[ domReference ].appendChild( node );
+			}
+		)
+	);
+};
+
+/**
+ * Performs a blocks preview using an iFrame.
+ *
+ * @param {object} props component's props
+ * @param {object} props.className CSS class to apply to component
+ * @param {string} props.bodyClassName CSS class to apply to the iframe's `<body>` tag
+ * @param {number} props.viewportWidth pixel width of the viewable size of the preview
+ * @param {Array}  props.blocks array of Gutenberg Block objects
+ * @param {object} props.settings block Editor settings object
+ */
+const BlockFramePreview = ( {
+	className = 'block-iframe-preview',
+	bodyClassName = 'block-iframe-preview-body',
+	viewportWidth,
+	blocks,
+	settings,
+} ) => {
+	const framePreviewRef = useRef();
+
+	// Set the initial scale factor.
+	const [ style, setStyle ] = useState( {
+		transform: `scale( 1 )`,
+	} );
+
+	// Rendering blocks list.
+	const renderedBlocks = useMemo( () => castArray( blocks ), [ blocks ] );
+	const [ recomputeBlockListKey, triggerRecomputeBlockList ] = useReducer( state => state + 1, 0 );
+	useLayoutEffect( triggerRecomputeBlockList, [ blocks ] );
+
+	/**
+	 * This function re scales the viewport depending on
+	 * the wrapper and the iframe width.
+	 */
+	const reScale = useCallback( () => {
+		const parentNode = get( framePreviewRef, [ 'current', 'parentNode' ] );
+		if ( ! parentNode ) {
+			return;
+		}
+
+		// Scaling iFrame.
+		const width = viewportWidth || framePreviewRef.current.offsetWidth;
+		const scale = parentNode.offsetWidth / viewportWidth;
+		const height = parentNode.offsetHeight / scale;
+
+		setStyle( {
+			width,
+			height,
+			transform: `scale( ${ scale } )`,
+		} );
+	}, [] );
+
+	// Populate iFrame styles.
+	useEffect( () => {
+		const head = get( framePreviewRef, [
+			'current',
+			'firstElementChild',
+			'contentDocument',
+			'head',
+		] );
+		const body = get( framePreviewRef, [
+			'current',
+			'firstElementChild',
+			'contentDocument',
+			'body',
+		] );
+
+		body.className = `${ bodyClassName } editor-styles-wrapper`;
+		loadStyles( { head, body } );
+		reScale();
+	}, [] );
+
+	// Scroll top the preview once it's rendered.
+	useEffect( () => {
+		const body = get( framePreviewRef, [
+			'current',
+			'firstElementChild',
+			'contentDocument',
+			'body',
+		] );
+		if ( ! body ) {
+			return;
+		}
+
+		// scroll to top when blocks changes.
+		body.scrollTop = 0;
+
+		const iFrameDocument = get( framePreviewRef, [
+			'current',
+			'firstElementChild',
+			'contentDocument',
+		] );
+		// const iFrameRenderedBlocksDOM = iFrameDocument.getElementById( 'rendered-blocks' );
+		// if ( iFrameRenderedBlocksDOM ) {
+		// 	iFrameDocument.body.removeChild( iFrameRenderedBlocksDOM );
+		// }
+
+		const renderedBlocksDOM = get( framePreviewRef, [ 'current', 'children' ] )[ 1 ];
```

Comment:
This seems potentially brittle. Should we just add a ref to the specific DOM node and have done with it?

Reply:
Fixed.